### ISSUE TITLE
Unexpected typing reports when using a subclass that overrides __iter__ in a incompatible manner

### ISSUE THREAD

**Issue by Questionsdenombres** [microsoft/pyright#3966]
## Environment data
- Windows-10-10.0.19042-SP0
- Python 3.10.7 (tags/v3.10.7:6cc6b13, Sep  5 2022, 14:08:36) - from python.org
- Visual Studio Code: 1.71.2
- Pylance v2022.9.20 (as a VScode extension)

## Code Snippet
```python
from collections.abc import Iterator
from itertools import chain

# --- Example microsoft/pylance-release#1 ---
class C(tuple[int, ...]):
    def __iter__(self) -> Iterator[str]: # type: ignore # (reportIncompatibleMethodOverride)
        yield from (str(x) for x in super().__iter__())
        return

C0 = C(tuple(range(1, 5)))
c1: list[str] = [x for x in C0 if x] # ok
# unnecessary 'if x' removes pylint report: "Unnecessary use of a comprehension, use list(C0) instead".
c2: list[str] = list(C0) # Expression of type "list[int]" cannot be assigned to declared type "list[str]"
c3: list[str] = [x for (_, x) in enumerate(C0)] # idem
c4: list[str] = [x for (_, x) in zip(range(1, 5), C(tuple(range(1, 5))))] # idem

# --- Example microsoft/pylance-release#2 ---
class DRange(tuple[range, ...]):
    def __iter__(self) -> Iterator[int]:  # type: ignore
        yield from chain(*super().__iter__())
        return

D0 = DRange((range(1, 3), range(7, 10)))
d1: list[int] = [x for x in D0 if x] # -> [1, 2, 7, 8, 9] # ok
d2: list[int] = list(D0) # Expression of type "list[range]" cannot be assigned to declared type "list[int]"
```

## Actual and Expected behavior

The first example is as simpler as possible.
`C` class is a subclass of `tuple[int, ...]`.
The superclass dunder `__iter__` method is an `Iterator[int]`.
It is is overridden in `C` as an `Iterator[str]`: items from the the superclass `__iter__` are stringified.

pylance then correctly warns and reports:
```text
Method "__iter__" overrides class "tuple" in an incompatible manner
Return type mismatch: base method returns type "Iterator[int]", override returns type "Iterator[str]"
```
I ignore the warning adding `# type: ignore` because this is what I want.

I was then expected that all iterations on `C` instances be typed as iterations of strings.
However the type checker seems to sometimes use the superclass `__iter__` type: `Iterator[int]`.
Typing errors are reported for the lines where `c2`, `c3` and `c4` are declared.
Python execution is correct.

The second example shows how I came to this: building a `DRange` class where `DRange` means "discontinuous range".
An instance shall be initialized with a tuple of disjoint and sorted ranges.
A typing error is reported for the line where `d2` is declared.
Python execution is correct too.

I don't know if this is a bug, but it is surprising.
I already hear the recommendation: *"Don't override a method in a incompatible manner!"*
Workarounds are easy. E.g:
- Create a dedicated method when you need to iterate differently from the superclass.
- The drawback is the lost of the transparent implicit use of the dunder method.

Thank you for the great pylance/pyright product.
Patrick - Paris

**Comment by erictraut** [comment]
The behavior you're seeing is intended, not a bug.

When pyright (the type checker that underlies pylance) needs to determine whether a class matches a protocol, such as `Iterator`, it first determines whether the class explicitly derives from that protocol — i.e. whether the protocol is part of its class hierarchy. If so, it doesn't need to do any further work. If a class doesn't explicitly derive from the protocol class, pyright then performs a more extensive (and costly) check to determine whether the class implements all of the methods and attributes that comprise the protocol. The latter technique is much more expensive, so it is used only when required.

In your code sample, the class `C` explicitly derives from `Iterator[int]`, so pyright doesn't bother performing any further checks when determining whether `C` is compatible with `Iterator[T]`. It can already see from the class hierarchy that `C` explicitly derives from `Iterator[int]`. The problem is that your class is violating the protocol that it claims to derive from by overriding `__iter__` and returning a value that is incompatible. Pyright detects this violation, but you have suppressed the error using a `# type: ignore`. Keep in mind that `# type: ignore` does not fix a type violation. It simply suppresses any errors that are on that line.

Here are some options you have:
1. If `C` is an iterable that returns `str` values, indicate that `C` derives from `tuple[str, ...]` rather than `tuple[int, ...]`.
2. If `C` acts like a an iterable that returns `int` values in some cases but `str` values in other cases, consider adding a method `C.as_strs` that returns a separate `Iterator[str]` object. Then you can use `list[C0.as_strs()]` without introducing any type violations.
